ink/lib.rs

```diff
@@ -1,44 +1,68 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use ink_env::chain_extension::FromStatusCode;
-use ink_lang as ink;
-use scale::Decode;
-pub use xcm::{VersionedMultiAsset, VersionedMultiLocation, VersionedResponse, VersionedXcm};
+use ink::env::Environment;
+pub use xcm::{
+    v3::prelude, VersionedMultiAsset, VersionedMultiLocation, VersionedResponse, VersionedXcm,
+};
 
-#[derive(Decode)]
-pub enum Error {
-    NoResponse = 1,
+/// The error type returned by the XCM extension.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum XCMError {
+    /// The XCM query requested via [`XCMExtension::take_response`] was not found
+    QueryNotFound = 1,
 }
 
-impl FromStatusCode for Error {
+impl ink::env::chain_extension::FromStatusCode for XCMError {
     fn from_status_code(status_code: u32) -> Result<(), Self> {
         match status_code {
             0 => Ok(()),
-            1 => Err(Self::NoResponse),
+            1 => Err(Self::QueryNotFound),
             _ => panic!("Unknown status code"),
         }
     }
 }
 
-#[ink::chain_extension]
-pub trait Extension {
-    type ErrorCode = Error;
-
-    #[ink(extension = 0x00010000, handle_status = false, returns_result = false)]
-    fn prepare_execute(xcm: VersionedXcm<()>) -> u64;
+type DefaultBlockNumber = <ink::env::DefaultEnvironment as Environment>::BlockNumber;
 
-    #[ink(extension = 0x00010001, handle_status = false, returns_result = false)]
-    fn execute();
+#[ink::chain_extension]
+pub trait XCMExtension {
+    type ErrorCode = XCMError;
 
-    #[ink(extension = 0x00010002, handle_status = false, returns_result = false)]
-    fn prepare_send(dest: VersionedMultiLocation, xcm: VersionedXcm<()>) -> VersionedMultiAsset;
+    /// Execute an XCM message locally, using the contract's address as the origin.
+    #[ink(extension = 1)]
+    fn execute(xcm: VersionedXcm<()>);
 
-    #[ink(extension = 0x00010003, handle_status = false, returns_result = false)]
-    fn send();
+    /// Send an XCM message from the contract to the specified destination.
+    #[ink(extension = 2)]
+    fn send(dest: VersionedMultiLocation, message: VersionedXcm<()>);
 
-    #[ink(extension = 0x00010004, handle_status = false, returns_result = false)]
-    fn new_query() -> u64;
+    /// Create a new query, using the contract's address as the responder.
+    ///
+    /// Returns the query id.
+    /// TODO: timeout should be generic over the BlockNumber but chain_extension does not support generics.
+    #[ink(extension = 3)]
+    fn new_query(timeout: DefaultBlockNumber, match_querier: VersionedMultiLocation) -> u64;
```

Comment:
not sure what's our best option here to pass the timeout, since we can't use generics...